### PR TITLE
perpetual: new module for mapping value types into memory

### PR DESCRIPTION
**Pull request by sdegtiarev**
I suggests a new module for mapping  files into memory. Instead of viewing the file as array of bytes, it directly maps the object onto file allowing to handle it as regular in-memory object. The data remain persistent after the program exits and may be re-opened and used again or shared between processes.
Dynamic arrays are also supported with some limitations.

### CODE REVIEW

**Comment by quickfur** [comment]
How is this different from `std.mmfile`?

**Comment by sdegtiarev** [comment]
This one offers different level of interface, while std.mmfile returns an character array modeling memory heap, std.perpetual creates an object. The difference is similar to one between malloc() and operator new in C++.
--------------------------------------------
On Tue, 8/18/15, H. S. Teoh <notifications@github.com> wrote:

 Subject: Re: [phobos] perpetual: new module for mapping value types into memory (#3562)
 To: "D-Programming-Language/phobos" <phobos@noreply.github.com>
 Cc: "sdegtiarev" <sdegtiarev@yahoo.com>
 Date: Tuesday, August 18, 2015, 2:09 PM
 
 How is this different from
 std.mmfile?
 
 —
 Reply to this email directly or view
 it on GitHub.

File: std/experimental/perpetual/perpetual.d [hunk 1]
```diff
@@ -0,0 +1,352 @@
...
+	// Perpetual!(immutable(short[]))  : [1, 0, 3, 0, 5, 0, 7, 0, 0, 0]
+	// Perpetual!(Color)               : green
+	// Perpetual!(char[])              : ABCDE___________________________
+	// Perpetual!string                : ABCDE___________________________
+	// Perpetual!(char[3][5])          : ["one", "...", "two", "...", "..."]
+	// Perpetual!(const(char[]))       : one...two......
+	// Perpetual!(char[3][])           : ["one", "...", "two", "...", "..."]
+
+
+	void main(string[] arg)
```
Comment (DmitryOlshansky) [inline std/experimental/perpetual/perpetual.d]: No main please. Use unittest. See also std.file unittests for `deleteme` and other ways to unittest with files

File: std/experimental/perpetual/perpetual.d [hunk 2]
```diff
@@ -0,0 +1,352 @@
...
+		} else
+			static assert(0);
+	}
+	
+	
+	
+	
+	
+	
+	private bool map(string path) {
```
Comment (DmitryOlshansky) [inline std/experimental/perpetual/perpetual.d]: Phobos style is brace on its own line. Yeah, I don't like it either but it's the way things are.

File: std/experimental/perpetual/perpetual.d [hunk 3]
```diff
@@ -0,0 +1,352 @@
...
+	~this() {
+		unmap();
+	}
+
+
+
+/**
+ * Flash memory to disc.
+ */
+	void sync() {
```
Comment (DmitryOlshansky) [inline std/experimental/perpetual/perpetual.d]: May want to expose some of version(Windows)/version(Posix) functions as stand-alone portable functions that pass around some opaque struct that describes the mapping. Then std.mmfile would use them too and the Q about reusing std.mmfile would be addressed.

File: std/experimental/perpetual/perpetual.d [hunk 4]
```diff
@@ -0,0 +1,352 @@
...
+} else {
+	static assert(0);
+}
+
+
+
+/**
+ * Persistently maps value type object to file
+ */
+class Perpetual(T)
```
Comment (DmitryOlshansky) [inline std/experimental/perpetual/perpetual.d]: Should be struct with either ref-counting or disabled this. There is no way to duplicate mapping

**Comment by sdegtiarev** [comment]
Oops, found something interesting (this is about zero-size file unmapping again)
Actually, munmap() is called in two places: destructor and ensureMapped(), which is called for every opSlice and, in case the file is empty, always tries to remap the memory.
So, the first error is thrown from munmap() from ensureMapped(), and another one during stack unwind  from inside munmap() from ~this();
The first one is treated as regular exception and the second one produces the core.exception.InvalidMemoryOperationError@(0)
The code:
unittest {
try {
	{ File(deleteme,"w"); } // create zero size file
	auto m=new MmFile(deleteme, MmFile.Mode.read, 0, null);
	assert(m.length == 0);
	m[0]; //ensureMapped() =>  index 0 is not mapped => munmap(data, 0) => throw

	writeln("done");
	remove(deleteme);

}
 // ~this() => munmap(data, 0) => throw again
catch(Exception x)
{
	writeln("error");
}
neither of writeln() statements  is ever reached

**Comment by quickfur** [comment]
Since this PR adds a whole new module to Phobos (albeit in `std.experimental`), it's probably a good idea to solicit for reviews on the forum. It should probably also go through the Phobos review process before being considered for merging.

**Pull request closed without merging**